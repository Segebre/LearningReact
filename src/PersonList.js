import React from 'react';
import axios from 'axios';


class PersonList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            people: [],
            isLoaded: false,
            error: null
        }
    }

    componentDidMount() {
        axios.get('https://jsonplaceholder.typicode.com/users')
        .then(response => {
            this.setState({isLoaded: true, people: response.data})
        }, error => {
            this.setState({isLoaded: false, error: error.message})
        })
    }

    render() {
        const isLoaded = this.state.isLoaded;
        const people = this.state.people;
        const error = this.state.error;

        if (error) {
            return <p>We found the folowing error: {error}</p>
        } else if (!isLoaded) {
            return <p>The site is loading...</p>
        } else {
            return (
                <ul>
                    {people.map((person) => (
                        <li>{person.name}</li>
                    ))}
                </ul>
            );
        }

    }
}

export default PersonList;