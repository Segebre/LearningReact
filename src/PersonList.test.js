import React from 'react';
import Renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import PersonList from './PersonList'
import mockAxios from 'axios';

jest.mock('axios');
mockAxios.get.mockImplementation(async () => {});

let container = null;
beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
})
afterEach(() => {
    ReactDOM.unmountComponentAtNode(container);
    container.remove();
    container = null;
});

xit('Renders without throwing', () => {
    let component;
    Renderer.act(() => {
        component = Renderer.create(<PersonList />);
    });
    const componentJson = component.toJSON();

    expect(componentJson).toMatchSnapshot();
});

xit("displays loading message", () => {
    const fakePeople = [
        { name: 'John' },
        { name: 'Mary' },
        { name: 'Monica' }
    ]

    mockAxios.get.mockResolvedValueOnce(fakePeople);
    
    act(() => {
        ReactDOM.render(<PersonList />, container);
    });

    expect(container.textContent).toBe('The site is loading...');
});

it("displays error message", async () => {
    const errorMessage = 'We found the folowing error: 404 not found.';

    mockAxios.get.mockRejectedValueOnce(new Error(errorMessage));
    
    act(() => {
        ReactDOM.render(<PersonList />, container);
    });

    await expect(() => mockAxios.get('')).toThrow('404 not found.');
});

// Tests
// https://jestjs.io/docs/en/mock-function-api#mockfnmockrejectedvalueoncevalue
// https://jestjs.io/docs/en/expect.html#tothrowerror

// Mocks
// https://jestjs.io/docs/en/expect.html#tothrowerror
// https://jestjs.io/docs/en/mock-functions.html#mocking-modules