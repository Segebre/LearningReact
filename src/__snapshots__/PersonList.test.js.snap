// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without throwing 1`] = `
<p>
  The site is loading...
</p>
`;
